File: packages/expo-perf-recorder/ios/PerfRecorderModule.swift
import ExpoModulesCore
import Foundation

// Native owner of persistence, per-dump analysis and rolling stats. The JS
// capture loop only serialises a V5 profiling export and hands it to
// `ingestDump`; everything below runs on a private serial queue, off the JS
// thread. In-memory rolling state survives Fast Refresh (the native runtime is
// not restarted on a JS reload); raw dumps are also ring-buffered to disk.
public class PerfRecorderModule: Module {
  private let queue = DispatchQueue(label: "com.artshin.perfrecorder.analyze")
  private let topN = 15

  // Rolling state (guarded by `queue`).
  private var maxDumps = 50
  private var dumpCount = 0
  private var totalCommits = 0
  private var totalReactCommitMs = 0.0
  private var rollingByName: [String: (ms: Double, count: Int)] = [:]       // actual
  private var rollingSelfByName: [String: (ms: Double, count: Int)] = [:]    // self-time
  private var rollingTriggers: [String: (times: Int, ms: Double, cascade: Int)] = [:]
  private var lastDump: [String: Any]?

  public func definition() -> ModuleDefinition {
    Name("PerfRecorder")

    AsyncFunction("configure") { (maxDumps: Int, promise: Promise) in
      self.queue.async {
        self.maxDumps = max(1, maxDumps)
        promise.resolve(nil)
      }
    }

    AsyncFunction("ingestDump") { (json: String, promise: Promise) in
      self.queue.async {
        self.ingest(json)
        promise.resolve(nil)
      }
    }

    AsyncFunction("getStats") { (promise: Promise) in
      self.queue.async {
        promise.resolve(self.statsJSON())
      }
    }

    AsyncFunction("clear") { (promise: Promise) in
      self.queue.async {
        self.dumpCount = 0
        self.totalCommits = 0
        self.totalReactCommitMs = 0
        self.rollingByName.removeAll()
        self.rollingSelfByName.removeAll()
        self.rollingTriggers.removeAll()
        self.lastDump = nil
        self.clearRing()
        promise.resolve(nil)
      }
    }

    AsyncFunction("listDumps") { (promise: Promise) in
      self.queue.async {
        promise.resolve(self.listDumpsJSON())
      }
    }

    AsyncFunction("getDumpSummary") { (name: String, promise: Promise) in
      self.queue.async {
        promise.resolve(self.dumpSummaryJSON(name))
      }
    }
  }

  // MARK: - Analysis

  private struct Analysis {
    let summary: [String: Any]
    let aggById: [Int: (ms: Double, count: Int)] // actual (subtree)
    let selfById: [Int: (ms: Double, count: Int)] // self-time
    let triggers: [String: (times: Int, ms: Double, cascade: Int)] // updater → cascade
    let nameById: [Int: String]
    let commits: Int
    let reactMs: Double
  }

  // Pure analysis of a V5 export's top-level object. No state mutation, so it is
  // reused by both ingest() and getDumpSummary() (re-analysing a persisted dump).
  private func analyze(_ top: [String: Any]) -> Analysis? {
    guard let roots = top["dataForRoots"] as? [[String: Any]] else { return nil }

    var commits = 0
    var fiberRecords = 0
    var reactMs = 0.0
    var tMin = Double.greatestFiniteMagnitude
    var tMax = -Double.greatestFiniteMagnitude
    var nameById: [Int: String] = [:]
    var aggById: [Int: (ms: Double, count: Int)] = [:]
    var selfById: [Int: (ms: Double, count: Int)] = [:]
    var triggers: [String: (times: Int, ms: Double, cascade: Int)] = [:]

    for r in roots {
      if let snaps = r["snapshots"] as? [String: Any] {
        for (k, v) in snaps {
          if let m = v as? [String: Any],
             let dn = m["displayName"] as? String,
             let id = Int(k) {
            nameById[id] = dn
          }
        }
      }
      guard let commitData = r["commitData"] as? [[String: Any]] else { continue }
      for com in commitData {
        commits += 1
        let dur = (com["duration"] as? NSNumber)?.doubleValue ?? 0
        reactMs += dur
        if let ts = (com["timestamp"] as? NSNumber)?.doubleValue {
          tMin = min(tMin, ts)
          tMax = max(tMax, ts)
        }
        let cascade = (com["fiberActualDurations"] as? [[Any]])?.count ?? 0
        if let updaters = com["updaters"] as? [[String: Any]] {
          // Distinct updater names per commit, each credited the commit cost —
          // "why this commit happened" (the trigger view).
          var seen = Set<String>()
          for u in updaters {
            if let id = (u["id"] as? NSNumber)?.intValue,
               let dn = u["displayName"] as? String {
              nameById[id] = dn
              if seen.insert(dn).inserted {
                var t = triggers[dn] ?? (times: 0, ms: 0, cascade: 0)
                t.times += 1
                t.ms += dur
                t.cascade += cascade
                triggers[dn] = t
              }
            }
          }
        }
        if let fad = com["fiberActualDurations"] as? [[Any]] {
          for pair in fad where pair.count >= 2 {
            guard
              let id = (pair[0] as? NSNumber)?.intValue,
              let ms = (pair[1] as? NSNumber)?.doubleValue
            else { continue }
            var cur = aggById[id] ?? (ms: 0, count: 0)
            cur.ms += ms
            cur.count += 1
            aggById[id] = cur
            fiberRecords += 1
          }
        }
        if let fsd = com["fiberSelfDurations"] as? [[Any]] {
          for pair in fsd where pair.count >= 2 {
            guard
              let id = (pair[0] as? NSNumber)?.intValue,
              let ms = (pair[1] as? NSNumber)?.doubleValue
            else { continue }
            var cur = selfById[id] ?? (ms: 0, count: 0)
            cur.ms += ms
            cur.count += 1
            selfById[id] = cur
          }
        }
      }
    }

    let wallSec = tMax > tMin ? (tMax - tMin) / 1000.0 : 0

    let summary: [String: Any] = [
      "ingestedAt": Date().timeIntervalSince1970 * 1000,
      "roots": roots.count,
      "commits": commits,
      "fiberRecords": fiberRecords,
      "reactCommitMs": reactMs,
      "wallSec": wallSec,
      "topFibers": topFiberList(aggById, nameById), // subtree (actual)
      "topBySelf": topFiberList(selfById, nameById), // self-time ("what's expensive")
      "triggers": triggerList(triggers), // updaters ("why it renders")
    ]
    return Analysis(
      summary: summary, aggById: aggById, selfById: selfById, triggers: triggers,
      nameById: nameById, commits: commits, reactMs: reactMs
    )
  }

  // Top-N named fibers from an id→(ms,count) map, descending by ms.
  private func topFiberList(
    _ agg: [Int: (ms: Double, count: Int)], _ nameById: [Int: String]
  ) -> [[String: Any]] {
    return agg
      .map { (id, v) in (nameById[id] ?? "(unnamed)", v.ms, v.count) }
      .sorted { $0.1 > $1.1 }
      .prefix(topN)
      .map { fiberDict(name: $0.0, ms: $0.1, count: $0.2) }
  }

  private func triggerList(
    _ triggers: [String: (times: Int, ms: Double, cascade: Int)]
  ) -> [[String: Any]] {
    return triggers
      .sorted { $0.value.ms > $1.value.ms }
      .prefix(topN)
      .map { triggerDict(name: $0.key, t: $0.value) }
  }

  private func ingest(_ json: String) {
    guard
      let data = json.data(using: .utf8),
      let top = (try? JSONSerialization.jsonObject(with: data)) as? [String: Any],
      let a = analyze(top)
    else { return }

    // Fold into rolling aggregates, keyed by NAME (ids reset on every reload).
    for (id, v) in a.aggById {
      let name = a.nameById[id] ?? "(unnamed)"
      var cur = rollingByName[name] ?? (ms: 0, count: 0)
      cur.ms += v.ms
      cur.count += v.count
      rollingByName[name] = cur
    }
    for (id, v) in a.selfById {
      let name = a.nameById[id] ?? "(unnamed)"
      var cur = rollingSelfByName[name] ?? (ms: 0, count: 0)
      cur.ms += v.ms
      cur.count += v.count
      rollingSelfByName[name] = cur
    }
    for (name, t) in a.triggers {
      var cur = rollingTriggers[name] ?? (times: 0, ms: 0, cascade: 0)
      cur.times += t.times
      cur.ms += t.ms
      cur.cascade += t.cascade
      rollingTriggers[name] = cur
    }

    dumpCount += 1
    totalCommits += a.commits
    totalReactCommitMs += a.reactMs
    lastDump = a.summary

    persistRaw(json)
  }

  private func fiberDict(name: String, ms: Double, count: Int) -> [String: Any] {
    [
      "name": name,
      "cumulativeMs": ms,
      "renderCount": count,
      "avgMs": count > 0 ? ms / Double(count) : 0,
    ]
  }

  private func triggerDict(
    name: String, t: (times: Int, ms: Double, cascade: Int)
  ) -> [String: Any] {
    [
      "name": name,
      "timesTriggered": t.times,
      "totalCommitMs": t.ms,
      "avgCascadeFibers": t.times > 0 ? Double(t.cascade) / Double(t.times) : 0,
    ]
  }

  private func rollingFiberList(
    _ rolling: [String: (ms: Double, count: Int)]
  ) -> [[String: Any]] {
    return rolling
      .sorted { $0.value.ms > $1.value.ms }
      .prefix(topN)
      .map { fiberDict(name: $0.key, ms: $0.value.ms, count: $0.value.count) }
  }

  private func statsJSON() -> String {
    let triggers = rollingTriggers
      .sorted { $0.value.ms > $1.value.ms }
      .prefix(topN)
      .map { triggerDict(name: $0.key, t: $0.value) }

    let stats: [String: Any] = [
      "dumpCount": dumpCount,
      "totalCommits": totalCommits,
      "totalReactCommitMs": totalReactCommitMs,
      "topFibers": rollingFiberList(rollingByName), // subtree (actual)
      "topBySelf": rollingFiberList(rollingSelfByName), // self-time
      "triggers": Array(triggers), // why renders happen
      "lastDump": lastDump as Any,
    ]
    guard
      let data = try? JSONSerialization.data(withJSONObject: stats),
      let str = String(data: data, encoding: .utf8)
    else { return "{}" }
    return str
  }

  // MARK: - Disk ring buffer

  private func ringDir() -> URL? {
    guard let caches = FileManager.default.urls(
      for: .cachesDirectory, in: .userDomainMask
    ).first else { return nil }
    let dir = caches.appendingPathComponent("perf-recorder", isDirectory: true)
    try? FileManager.default.createDirectory(
      at: dir, withIntermediateDirectories: true
    )
    return dir
  }

  private func persistRaw(_ json: String) {
    guard let dir = ringDir() else { return }
    let file = dir.appendingPathComponent("dump-\(Int(Date().timeIntervalSince1970 * 1000)).json")
    try? json.data(using: .utf8)?.write(to: file)
    pruneRing(dir)
  }

  private func pruneRing(_ dir: URL) {
    guard let files = try? FileManager.default.contentsOfDirectory(
      at: dir, includingPropertiesForKeys: nil
    ).filter({ $0.lastPathComponent.hasPrefix("dump-") }) else { return }
    if files.count <= maxDumps { return }
    let sorted = files.sorted { $0.lastPathComponent < $1.lastPathComponent }
    for url in sorted.prefix(files.count - maxDumps) {
      try? FileManager.default.removeItem(at: url)
    }
  }

  private func clearRing() {
    guard let dir = ringDir() else { return }
    try? FileManager.default.removeItem(at: dir)
  }

  // Reject anything that isn't a bare `dump-<digits>.json` basename — guards
  // getDumpSummary against path traversal out of the ring dir.
  private func isValidDumpName(_ name: String) -> Bool {
    return !name.contains("/") && !name.contains("..")
      && name.hasPrefix("dump-") && name.hasSuffix(".json")
  }

  private func timestampFromName(_ name: String) -> Double {
    let trimmed = name.dropFirst("dump-".count).dropLast(".json".count)
    return Double(trimmed) ?? 0
  }

  private func listDumpsJSON() -> String {
    guard
      let dir = ringDir(),
      let urls = try? FileManager.default.contentsOfDirectory(
        at: dir, includingPropertiesForKeys: [.fileSizeKey]
      )
    else { return "[]" }
    let entries: [[String: Any]] = urls
      .filter { $0.lastPathComponent.hasPrefix("dump-") }
      .sorted { $0.lastPathComponent > $1.lastPathComponent } // newest first
      .map { url in
        let size = (try? url.resourceValues(forKeys: [.fileSizeKey]).fileSize) ?? 0
        return [
          "name": url.lastPathComponent,
          "sizeBytes": size,
          "ingestedAt": timestampFromName(url.lastPathComponent),
        ]
      }
    guard
      let data = try? JSONSerialization.data(withJSONObject: entries),
      let str = String(data: data, encoding: .utf8)
    else { return "[]" }
    return str
  }

  private func dumpSummaryJSON(_ name: String) -> String {
    guard isValidDumpName(name), let dir = ringDir() else { return "null" }
    let file = dir.appendingPathComponent(name)
    guard
      let data = try? Data(contentsOf: file),
      let top = (try? JSONSerialization.jsonObject(with: data)) as? [String: Any],
      let a = analyze(top)
    else { return "null" }
    var summary = a.summary
    summary["ingestedAt"] = timestampFromName(name) // dump's own time, not now
    guard
      let out = try? JSONSerialization.data(withJSONObject: summary),
      let str = String(data: out, encoding: .utf8)
    else { return "null" }
    return str
  }
}

File: packages/expo-perf-recorder/android/src/main/java/expo/modules/perfrecorder/PerfRecorderModule.kt
package expo.modules.perfrecorder

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.Promise
import org.json.JSONArray
import org.json.JSONObject
import java.io.File
import java.util.concurrent.Executors

// Native owner of persistence, per-dump analysis and rolling stats. The JS
// capture loop only serialises a V5 profiling export and hands it to
// `ingestDump`; everything below runs on a single-threaded executor, off the JS
// thread. In-memory rolling state survives Fast Refresh (the native runtime is
// not restarted on a JS reload); raw dumps are also ring-buffered to disk.
class PerfRecorderModule : Module() {
  private val worker = Executors.newSingleThreadExecutor()
  private val topN = 15

  // Rolling state (only touched on `worker`).
  private var maxDumps = 50
  private var dumpCount = 0
  private var totalCommits = 0
  private var totalReactCommitMs = 0.0
  private val rollingByName = HashMap<String, DoubleArray>() // name -> [ms, count] (actual)
  private val rollingSelfByName = HashMap<String, DoubleArray>() // name -> [ms, count] (self)
  private val rollingTriggers = HashMap<String, DoubleArray>() // name -> [times, ms, cascade]
  private var lastDump: JSONObject? = null

  override fun definition() = ModuleDefinition {
    Name("PerfRecorder")

    AsyncFunction("configure") { value: Int, promise: Promise ->
      worker.execute {
        maxDumps = if (value < 1) 1 else value
        promise.resolve(null)
      }
    }

    AsyncFunction("ingestDump") { json: String, promise: Promise ->
      worker.execute {
        try {
          ingest(json)
        } catch (e: Exception) {
          // Never crash the dev app over a malformed dump.
        }
        promise.resolve(null)
      }
    }

    AsyncFunction("getStats") { promise: Promise ->
      worker.execute { promise.resolve(statsJson()) }
    }

    AsyncFunction("clear") { promise: Promise ->
      worker.execute {
        dumpCount = 0
        totalCommits = 0
        totalReactCommitMs = 0.0
        rollingByName.clear()
        rollingSelfByName.clear()
        rollingTriggers.clear()
        lastDump = null
        clearRing()
        promise.resolve(null)
      }
    }

    AsyncFunction("listDumps") { promise: Promise ->
      worker.execute { promise.resolve(listDumpsJson()) }
    }

    AsyncFunction("getDumpSummary") { name: String, promise: Promise ->
      worker.execute {
        val out = try {
          dumpSummaryJson(name)
        } catch (e: Exception) {
          "null"
        }
        promise.resolve(out)
      }
    }
  }

  // MARK: analysis

  private class Analysis(
    val summary: JSONObject,
    val aggMs: HashMap<Int, Double>,
    val aggCount: HashMap<Int, Int>,
    val selfMs: HashMap<Int, Double>,
    val selfCount: HashMap<Int, Int>,
    val triggers: HashMap<String, DoubleArray>, // name -> [times, ms, cascade]
    val nameById: HashMap<Int, String>,
    val commits: Int,
    val reactMs: Double,
  )

  // Pure analysis of a V5 export. No state mutation, so it is reused by both
  // ingest() and getDumpSummary() (re-analysing a persisted dump).
  private fun analyze(top: JSONObject): Analysis? {
    val roots = top.optJSONArray("dataForRoots") ?: return null

    var commits = 0
    var fiberRecords = 0
    var reactMs = 0.0
    var tMin = Double.MAX_VALUE
    var tMax = -Double.MAX_VALUE
    val nameById = HashMap<Int, String>()
    val aggMs = HashMap<Int, Double>()
    val aggCount = HashMap<Int, Int>()
    val selfMs = HashMap<Int, Double>()
    val selfCount = HashMap<Int, Int>()
    val triggers = HashMap<String, DoubleArray>() // name -> [times, ms, cascade]

    for (ri in 0 until roots.length()) {
      val r = roots.optJSONObject(ri) ?: continue

      r.optJSONObject("snapshots")?.let { snaps ->
        val keys = snaps.keys()
        while (keys.hasNext()) {
          val k = keys.next()
          val m = snaps.optJSONObject(k) ?: continue
          val dn = m.optString("displayName", "")
          val id = k.toIntOrNull()
          if (dn.isNotEmpty() && id != null) nameById[id] = dn
        }
      }

      val commitData = r.optJSONArray("commitData") ?: continue
      for (ci in 0 until commitData.length()) {
        val com = commitData.optJSONObject(ci) ?: continue
        commits++
        val dur = com.optDouble("duration", 0.0)
        reactMs += dur
        val ts = com.optDouble("timestamp", Double.NaN)
        if (!ts.isNaN()) {
          if (ts < tMin) tMin = ts
          if (ts > tMax) tMax = ts
        }
        val cascade = com.optJSONArray("fiberActualDurations")?.length() ?: 0
        com.optJSONArray("updaters")?.let { updaters ->
          // Distinct updater names per commit, each credited the commit cost.
          val seen = HashSet<String>()
          for (ui in 0 until updaters.length()) {
            val u = updaters.optJSONObject(ui) ?: continue
            val id = u.optInt("id", -1)
            val dn = u.optString("displayName", "")
            if (id >= 0 && dn.isNotEmpty()) nameById[id] = dn
            if (dn.isNotEmpty() && seen.add(dn)) {
              val t = triggers.getOrPut(dn) { doubleArrayOf(0.0, 0.0, 0.0) }
              t[0] += 1
              t[1] += dur
              t[2] += cascade
            }
          }
        }
        com.optJSONArray("fiberActualDurations")?.let { fad ->
          for (fi in 0 until fad.length()) {
            val pair = fad.optJSONArray(fi) ?: continue
            if (pair.length() < 2) continue
            val id = pair.optInt(0, -1)
            val ms = pair.optDouble(1, Double.NaN)
            if (id < 0 || ms.isNaN()) continue
            aggMs[id] = (aggMs[id] ?: 0.0) + ms
            aggCount[id] = (aggCount[id] ?: 0) + 1
            fiberRecords++
          }
        }
        com.optJSONArray("fiberSelfDurations")?.let { fsd ->
          for (fi in 0 until fsd.length()) {
            val pair = fsd.optJSONArray(fi) ?: continue
            if (pair.length() < 2) continue
            val id = pair.optInt(0, -1)
            val ms = pair.optDouble(1, Double.NaN)
            if (id < 0 || ms.isNaN()) continue
            selfMs[id] = (selfMs[id] ?: 0.0) + ms
            selfCount[id] = (selfCount[id] ?: 0) + 1
          }
        }
      }
    }

    val wallSec = if (tMax > tMin) (tMax - tMin) / 1000.0 else 0.0

    val summary = JSONObject().apply {
      put("ingestedAt", System.currentTimeMillis())
      put("roots", roots.length())
      put("commits", commits)
      put("fiberRecords", fiberRecords)
      put("reactCommitMs", reactMs)
      put("wallSec", wallSec)
      put("topFibers", fiberListFromIds(aggMs, aggCount, nameById)) // subtree (actual)
      put("topBySelf", fiberListFromIds(selfMs, selfCount, nameById)) // self-time
      put("triggers", triggerListJson(triggers)) // why it renders
    }
    return Analysis(
      summary, aggMs, aggCount, selfMs, selfCount, triggers, nameById, commits, reactMs
    )
  }

  // Top-N named fibers from id→ms / id→count maps, descending by ms.
  private fun fiberListFromIds(
    ms: Map<Int, Double>,
    count: Map<Int, Int>,
    nameById: Map<Int, String>,
  ): JSONArray {
    val arr = JSONArray()
    ms.keys
      .map { id -> Triple(nameById[id] ?: "(unnamed)", ms[id] ?: 0.0, count[id] ?: 0) }
      .sortedByDescending { it.second }
      .take(topN)
      .forEach { arr.put(fiberObj(it.first, it.second, it.third)) }
    return arr
  }

  private fun triggerListJson(map: Map<String, DoubleArray>): JSONArray {
    val arr = JSONArray()
    map.entries
      .sortedByDescending { it.value[1] } // by totalCommitMs
      .take(topN)
      .forEach { arr.put(triggerObj(it.key, it.value[0].toInt(), it.value[1], it.value[2].toInt())) }
    return arr
  }

  private fun ingest(json: String) {
    val a = analyze(JSONObject(json)) ?: return

    // Fold into rolling aggregates, keyed by NAME (ids reset on every reload).
    for (id in a.aggMs.keys) {
      val name = a.nameById[id] ?: "(unnamed)"
      val cur = rollingByName.getOrPut(name) { doubleArrayOf(0.0, 0.0) }
      cur[0] += a.aggMs[id] ?: 0.0
      cur[1] += (a.aggCount[id] ?: 0).toDouble()
    }
    for (id in a.selfMs.keys) {
      val name = a.nameById[id] ?: "(unnamed)"
      val cur = rollingSelfByName.getOrPut(name) { doubleArrayOf(0.0, 0.0) }
      cur[0] += a.selfMs[id] ?: 0.0
      cur[1] += (a.selfCount[id] ?: 0).toDouble()
    }
    for ((name, t) in a.triggers) {
      val cur = rollingTriggers.getOrPut(name) { doubleArrayOf(0.0, 0.0, 0.0) }
      cur[0] += t[0]
      cur[1] += t[1]
      cur[2] += t[2]
    }

    dumpCount++
    totalCommits += a.commits
    totalReactCommitMs += a.reactMs
    lastDump = a.summary

    persistRaw(json)
  }

  private fun fiberObj(name: String, ms: Double, count: Int): JSONObject =
    JSONObject().apply {
      put("name", name)
      put("cumulativeMs", ms)
      put("renderCount", count)
      put("avgMs", if (count > 0) ms / count else 0.0)
    }

  private fun triggerObj(name: String, times: Int, ms: Double, cascade: Int): JSONObject =
    JSONObject().apply {
      put("name", name)
      put("timesTriggered", times)
      put("totalCommitMs", ms)
      put("avgCascadeFibers", if (times > 0) cascade.toDouble() / times else 0.0)
    }

  // Top-N from a rolling name->[ms, count] map, descending by ms.
  private fun rollingFiberList(map: Map<String, DoubleArray>): JSONArray {
    val arr = JSONArray()
    map.entries
      .sortedByDescending { it.value[0] }
      .take(topN)
      .forEach { arr.put(fiberObj(it.key, it.value[0], it.value[1].toInt())) }
    return arr
  }

  private fun statsJson(): String {
    return JSONObject().apply {
      put("dumpCount", dumpCount)
      put("totalCommits", totalCommits)
      put("totalReactCommitMs", totalReactCommitMs)
      put("topFibers", rollingFiberList(rollingByName)) // subtree (actual)
      put("topBySelf", rollingFiberList(rollingSelfByName)) // self-time
      put("triggers", triggerListJson(rollingTriggers)) // why renders happen
      put("lastDump", lastDump ?: JSONObject.NULL)
    }.toString()
  }

  // MARK: disk ring buffer

  private fun ringDir(): File? {
    val ctx = appContext.reactContext ?: return null
    val dir = File(ctx.cacheDir, "perf-recorder")
    if (!dir.exists()) dir.mkdirs()
    return dir
  }

  private fun persistRaw(json: String) {
    val dir = ringDir() ?: return
    File(dir, "dump-${System.currentTimeMillis()}.json").writeText(json)
    pruneRing(dir)
  }

  private fun pruneRing(dir: File) {
    val files = dir.listFiles { f -> f.name.startsWith("dump-") } ?: return
    if (files.size <= maxDumps) return
    files.sortedBy { it.name }
      .take(files.size - maxDumps)
      .forEach { it.delete() }
  }

  private fun clearRing() {
    ringDir()?.deleteRecursively()
  }

  // Reject anything that isn't a bare `dump-<digits>.json` basename — guards
  // getDumpSummary against path traversal out of the ring dir.
  private val dumpNameRe = Regex("^dump-\\d+\\.json$")

  private fun timestampFromName(name: String): Long =
    name.removePrefix("dump-").removeSuffix(".json").toLongOrNull() ?: 0L

  private fun listDumpsJson(): String {
    val dir = ringDir() ?: return "[]"
    val files = dir.listFiles { f -> f.name.startsWith("dump-") } ?: return "[]"
    val arr = JSONArray()
    files.sortedByDescending { it.name }.forEach { f -> // newest first
      arr.put(JSONObject().apply {
        put("name", f.name)
        put("sizeBytes", f.length())
        put("ingestedAt", timestampFromName(f.name))
      })
    }
    return arr.toString()
  }

  private fun dumpSummaryJson(name: String): String {
    if (!dumpNameRe.matches(name)) return "null"
    val dir = ringDir() ?: return "null"
    val f = File(dir, name)
    if (!f.exists()) return "null"
    val a = analyze(JSONObject(f.readText())) ?: return "null"
    a.summary.put("ingestedAt", timestampFromName(name)) // dump's own time
    return a.summary.toString()
  }
}
